sys/net/gnrc/link_layer/lwmac/lwmac.c

```diff
@@ -0,0 +1,745 @@
+        break;
+
+    /* Receiving incoming data */
+    case RECEIVING:
+        rtt_handler(LWMAC_EVENT_RTT_PAUSE); /* No duty cycling while TXing */
+        _set_netdev_state(&lwmac, NETOPT_STATE_IDLE);  /* Power up netdev */
+        break;
+
+    case STOPPED:
+        _set_netdev_state(&lwmac, NETOPT_STATE_OFF);
+        break;
+
+    /*********************** Control states ***********************************/
+    case START:
+        rtt_handler(LWMAC_EVENT_RTT_START);
+        lwmac_set_state(LISTENING);
+        break;
+
+    case STOP:
+        rtt_handler(LWMAC_EVENT_RTT_STOP);
+        lwmac_set_state(STOPPED);
+        break;
+
+    case RESET:
+        LOG_WARNING("Reset not yet implemented\n");
+        lwmac_set_state(STOP);
+        lwmac_set_state(START);
+        break;
+
+    /**************************************************************************/
+    default:
+        LOG_DEBUG("No actions for entering state %u\n", newstate);
+        return;
+    }
+
+    lwmac_schedule_update();
+}
+
+/******************************************************************************/
+
+/* Main state machine. Call whenever something happens */
+// TODO: Don't use global variables
+bool lwmac_update(void)
+{
+    lwmac.needs_rescheduling = false;
+
+    switch(lwmac.state)
+    {
+    case SLEEPING:
+
+        /* If a packet is scheduled, no other (possible earlier) packet can be
+         * sent before the first one is handled, even no broadcast
+         */
+        if( !lwmac_timeout_is_running(&lwmac, TIMEOUT_WAIT_FOR_DEST_WAKEUP) ) {
+
+            /* Check if there are broadcasts to send and transmit immediately */
+            if(packet_queue_length(&(lwmac.tx.neighbours[0].queue)) > 0) {
+                lwmac.tx.current_neighbour = &(lwmac.tx.neighbours[0]);
+                lwmac_set_state(TRANSMITTING);
+                break;
+            }
+
+            lwmac_tx_neighbour_t* neighbour = _next_tx_neighbour(&lwmac);
+
+            if(neighbour != NULL) {
+                /* Offset in microseconds when the earliest (phase) destination
+                 * node wakes up that we have packets for. */
+                int time_until_tx = RTT_TICKS_TO_US(_ticks_until_phase(neighbour->phase));
+
+                /* If there's not enough time to prepare a WR to catch the phase
+                 * postpone to next interval */
+                if (time_until_tx < LWMAC_WR_PREPARATION_US) {
+                    time_until_tx += LWMAC_WAKEUP_INTERVAL_US;
+                }
+
+                time_until_tx -= LWMAC_WR_PREPARATION_US;
+                lwmac_set_timeout(&lwmac, TIMEOUT_WAIT_FOR_DEST_WAKEUP, time_until_tx);
+
+                /* Register neighbour to be the next */
+                lwmac.tx.current_neighbour = neighbour;
+
+                /* Stop dutycycling, we're preparing to send. This prevents the
+                 * timeout arriving late, so that the destination phase would
+                 * be missed. */
```

Comment:
> 3) In case the CPU is now in the lpm sleep mode, a) the RTT module is independent and is still running, right?

Yes. 

 b) In case there are still only two threads, LwMAC and the idle thread, and the CPU is now in the sleep mode, when the RTT alarm is triggered/generated, RTT callback function (i.e., rtt_cb()) will be executed. 

Yes. Note: this callback function is called independent of any running thread (and not called in thread context). Think of it as if someone else, not your program, is calling it.

> But, how this RTT alarm will wakeup the CPU in the current LwMAC.

Not the callback itself wakes up the cpu. The hardware inside the CPU notices the alarm, wakes up the cpu and then calls the `rtt_cb()`. 

> Since in rtt_cb(), it only sends a message to LwMAC thread which is not running, and the rtt_handler() (which sets lpm_prevent_sleep |= LWMAC_LPM_MASK) is in the LwMAC thread that will only be called if LwMAC thread runs.

As I explained, LwMAC is still running. The message sending is neccessary, because the callback is not executed in thread context (see above) and nasty things can happen if you're not careful. After `rtt_cb()` has finished, execution will continue in the idle thread. Since we had sent a message to LwMAC, the scheduler delivers this message and LwMAC is running again.


If you want to read some background to OS design that is employed in RIOT, see (Cooperative Multitasking])[https://en.wikipedia.org/wiki/Cooperative_multitasking] and (Microkernel)[https://en.wikipedia.org/wiki/Microkernel]